Replace std thread hardware_concurrency by system calls.

dvicini opened the pull request:
This ensures that the number of used threads corresponds to the number of threads available to the process, instead of the number of physical CPU cores/threads.

This function is actually simply imported from Mitsuba 3, where it already exists in `util.cpp`. It probably makes sense to have this logic moved to nanothread, so the entire system uses a consistent number of threads throughout. If we merge this, we can replace the corresponding function in mitsuba's util.cpp

wjakob commented:
Can you confirm also that this fixes the issue on your end? Also, one small change request: let's keep on using the standard function (which didn't even exist back in the days) for Windows and macOS.

dvicini commented:
Yes, it fixes the issue on my end. Good point about using the standard function for Windows/MacOS, I changed it now. 

Once this is merged we can also remove the function in mitsuba's util.cpp and replace it by a call to nanothread::core_count()

wjakob commented on src/nanothread.cpp:
```diff
@@ -63,6 +65,65 @@ struct Worker {
+    /* Determine the number of present cores */
+    uint32_t ncores = sysconf(_SC_NPROCESSORS_CONF);
+
+    /* Don't try to query CPU affinity if running inside Valgrind */
```
Use `//` for single-line comments (here and elsewhere). (Those default conventions had changed over the years..)

wjakob commented on src/nanothread.cpp:
```diff
@@ -63,6 +65,65 @@ struct Worker {
+        for (uint32_t i = 0; i < 10; ++i) {
+            size = CPU_ALLOC_SIZE(ncores_logical);
+            cpuset = CPU_ALLOC(ncores_logical);
+            if (!cpuset) {
```
This function is exposed through a C-level API, and the default assumption there for some platforms like on Windows is that functions cannot throw an exception (see the `NANOTHREAD_THROW` macro that is used to annotate a few). My suggestion would be that this function is implemented to _never_ throw, and that it falls back to `std::thread::hardware_concurrency` in case of a failure.

Remove unnecessary braces, return `0` to signal a failure.

dvicini replied:
Yes makes sense, I wasn't convinced by the exceptions either and just using the hardware_concurrency as a default makes sense.

wjakob commented on src/nanothread.cpp:
```diff
@@ -63,6 +65,65 @@ struct Worker {
+            ncores_logical *= 2;
+        }
+
+        if (retval) {
```
ditto

wjakob commented on src/nanothread.cpp:
```diff
@@ -63,6 +65,65 @@ struct Worker {
+
+#if defined(__linux__)
+    /* Determine the number of present cores */
+    uint32_t ncores = sysconf(_SC_NPROCESSORS_CONF);
```
Is that the same as `std::thread::hardware_concurrency`?

dvicini replied:
I think so, again sysconf is the legacy way of getting this information it seems.

wjakob reviewed (changes requested):
A few more small things.

dvicini commented:
I removed exceptions and replaced them by using the default. I still added a printf to stderror. As far as I understand, these calls failing could still indicate an actual problem, so better for to not be completely silent

wjakob commented:
Thanks!

dvicini commented:
Ah you were fast on clicking merge, I found a minor issue that I now fixed on master (the cache variable was updated too early)

wjakob commented:
Thanks!
